sdk/tests/conformance2/textures/misc/tex-unpack-params-with-flip-y-and-premultiply-alpha.html

```diff
@@ -0,0 +1,503 @@
+  var bytesPerRow = actualWidth * bytesPerPixel + padding;
+  var bytesLastRow = bytesPerPixel * width;
+  var size = bytesPerRow * (height - 1) + bytesLastRow;
+  var skipSize = 0;
+  if (testCase.skipPixels > 0)
+    skipSize += bytesPerPixel * testCase.skipPixels;
+  if (testCase.skipRows > 0)
+    skipSize += bytesPerRow * testCase.skipRows;
+  return {size: size,
+          bytesPerRow: bytesPerRow,
+          bytesLastRow: bytesLastRow,
+          padding: padding,
+          skipSize: skipSize,
+          totalSize: size + skipSize};
+}
+
+function copyData(srcData, srcIndex, dstData, dstIndex, size, premultiply_alpha) {
+  for (var ii = 0; ii < size / 4; ++ii) {
+    var factor = 1.0;
+    if (premultiply_alpha)
+      factor = srcData[srcIndex + ii * 4 + 3] / 255.0;
+    dstData[dstIndex + ii * 4] = srcData[srcIndex + ii * 4] * factor;
+    dstData[dstIndex + ii * 4 + 1] = srcData[srcIndex + ii * 4 + 1] * factor;
+    dstData[dstIndex + ii * 4 + 2] = srcData[srcIndex + ii * 4 + 2] * factor;
+    dstData[dstIndex + ii * 4 + 3] = srcData[srcIndex + ii * 4 + 3];
+  }
+}
+
+function unpackPixels(srcData, width, height, imageSizes, flip_y, premultiply_alpha) {
+  var bytesPerPixel = 4;
+  var unpackedSize = width * height * bytesPerPixel;
+  var dstData = setupArrayBuffer(unpackedSize, false);
+  var srcIndex = imageSizes.skipSize;
+  var dstIndex = 0;
+  var inc = width * bytesPerPixel;
+  if (flip_y)
+    dstIndex += (height - 1) * inc;
+  for (var y = 0; y < height; ++y) {
+    copyData(srcData, srcIndex, dstData, dstIndex, width * bytesPerPixel, premultiply_alpha);
+    srcIndex += imageSizes.bytesPerRow;
+    if (flip_y)
+      dstIndex -= inc;
+    else
+      dstIndex += inc;
+  }
+  if (flip_y)
+    dstIndex += height * inc;
+  return dstData;
+}
+
+function getPixelsFromTexture2D(gl, tex, xoffset, yoffset, width, height) {
+  var fbo = gl.createFramebuffer();
+  gl.bindFramebuffer(gl.FRAMEBUFFER, fbo);
+  gl.framebufferTexture2D(gl.FRAMEBUFFER, gl.COLOR_ATTACHMENT0, gl.TEXTURE_2D, tex, 0);
+  var bytesPerReadbackPixel = 4;
+  var readbackBuffer = setupArrayBuffer(width * height * bytesPerReadbackPixel, false);
+  gl.readPixels(xoffset, yoffset, width, height, gl.RGBA, gl.UNSIGNED_BYTE, readbackBuffer);
+  wtu.glErrorShouldBe(gl, gl.NO_ERROR, "read back texture pixels should succeed");
+  gl.bindFramebuffer(gl.FRAMEBUFFER, null);
+  gl.deleteFramebuffer(fbo);
+  return readbackBuffer;
+}
+
+function comparePixels(buffer1, buffer2, tolerance) {
+  if (buffer1.length != buffer2.length || buffer1.length % 4 != 0) {
+    testFailed("compare pixels: invalid buffer size");
+    return;
+  }
+  var count = 0;
+  for (var ii = 0; ii < buffer1.length / 4; ++ii) {
+    var distance = 0;
+    for (var jj = 0; jj < 4; ++jj) {
+      var diff = buffer1[ii * 4 + jj] - buffer2[ii * 4 + jj];
+      distance += diff * diff;
+    }
+    if (Math.sqrt(diff) > tolerance) {
```

Comment:
Ooh, math! :D This seems fine; other tests have used absolute differences, but this is probably better.